modules/phase_field/src/action/MortarPeriodicAction.C

```diff
@@ -0,0 +1,152 @@
+  std::string periodicity_name = name();
+
+  // axis names
+  const std::vector<BoundaryName> axis = { "x", "y", "z" };
+
+  // boundaries
+  const std::vector<BoundaryName> boundary_names = { "left", "bottom", "back" };
+  // opposite boundaries
+  const std::vector<BoundaryName> opposite_boundary_names = { "right", "top", "front" };
+
+  // mortar subdomains
+  const std::vector<SubdomainID> & mortar_subdomains = mesh->getMortarSubdomains();
+
+  // iterate over the periodic directions
+  for (unsigned short i = 0; i < dim; ++i)
+    if (mesh->getPeriodicDirections().contains(i))
+    {
+      // initialize subdomain restriction set
+      std::set<SubdomainID> subdomain_restriction = { mortar_subdomains[i] };
+
+      // Lagrange multiplier variable base name
+      const std::string lm_base = "lm_" + periodicity_name + "_" + boundary_names[i];
+
+      // mortar interface name
+      const std::string mi_name = "mi_" + periodicity_name + "_" + boundary_names[i] + '_' + opposite_boundary_names[i];
+
+      //
+      // Add Lagrange multiplier variables
+      //
+
+      if (_current_task == "add_variable")
+      {
+        for (auto & var : _variables)
+        {
+          switch (_periodicity)
+          {
+            case 0: // gradient
+              for (unsigned int j = 0; j < dim; ++j)
+                _problem->addVariable(lm_base + "_" + var + "_d" + axis[j],
+                                      FEType(Utility::string_to_enum<Order>("FIRST"),
+                                             Utility::string_to_enum<FEFamily>("LAGRANGE")),
+                                      1.0, &subdomain_restriction);
+              break;
+
+            case 1: // value
+              _problem->addVariable(lm_base + "_" + var,
+                                    FEType(Utility::string_to_enum<Order>("FIRST"),
+                                           Utility::string_to_enum<FEFamily>("LAGRANGE")),
+                                    1.0, &subdomain_restriction);
+              break;
+
+            default:
+              mooseError("Periodicity type not implemented");
+          }
+        }
+      }
+
+      //
+      // Add Mortar interfaces
+      //
+
+      if (_current_task == "add_mortar_interface")
+      {
+        _mesh->addMortarInterface(mi_name, boundary_names[i], opposite_boundary_names[i], Moose::stringify(mortar_subdomains[i]));
+        if (_displaced_mesh)
+          _displaced_mesh->addMortarInterface(mi_name, boundary_names[i], opposite_boundary_names[i], Moose::stringify(mortar_subdomains[i]));
+      }
+
+      //
+      // Add Constraints
+      //
+
+      if (_current_task == "add_constraint")
+      {
+        for (auto & var : _variables)
+        {
+          const std::string ct_base = "ct_" + periodicity_name + "_" + boundary_names[i] + "_" + var;
+          switch (_periodicity)
+          {
+            case 0: // gradient
+              for (unsigned int j = 0; j < dim; ++j)
```

Comment:
> Nobody gets types correct

Grrr, `MultiMooseEnum::contains()` accepts arguments of type `int` or `unsigned short`. It fails to compile when passing in an `unsigned int`...